File: sdk/dart/test/conformance.dart
import 'dart:async';

import 'package:oresoftware_next_loggers/oresoftware_next_loggers.dart';

Future<void> main() async {
  final otel = <Map<String, Object?>>[];
  final supabase = <Map<String, Object?>>[];
  final logger = Logger(
    appName: 'payments',
    name: 'audit',
    fields: const {'environment': 'test'},
    idFactory: () => 'dart-record-1',
    clock: () => '2026-01-02T03:04:05.000Z',
    transports: <LogTransport>[
      OpenTelemetryTransport(otel.add),
      SupabaseTransport(supabase.add),
    ],
  );

  final record = await withLogContext(
    const LogContext(
      traceId: '0123456789abcdef0123456789abcdef',
      spanId: '0123456789abcdef',
      traceFlags: 1,
      traceState: 'vendor=value',
      fields: {'requestId': 'request-1'},
      tags: ['otel', 'flutter'],
    ),
    () => logger.error('payment failed', fields: const {'orderId': 'order-42'}),
  );

  assert(record['schema'] == nextLoggersSchema);
  assert(record['level'] == 'ERROR');
  assert(record['traceId'] == '0123456789abcdef0123456789abcdef');
  final fields = record['fields']! as Map<String, Object?>;
  assert(fields['otel.span_id'] == '0123456789abcdef');
  assert(fields['requestId'] == 'request-1');
  assert(fields['orderId'] == 'order-42');
  assert(otel.length == 1);
  assert(otel.single['severityNumber'] == 17);
  assert(supabase.length == 1);
  assert(currentLogContext == null);

  final traces = await Future.wait<String>(<Future<String>>[
    Future<String>(() async {
      return withLogContext(
        const LogContext(traceId: 'trace-a', spanId: 'span-a'),
        () async => (await logger.info('a'))['traceId']! as String,
      );
    }),
    Future<String>(() async {
      return withLogContext(
        const LogContext(traceId: 'trace-b', spanId: 'span-b'),
        () async => (await logger.info('b'))['traceId']! as String,
      );
    }),
  ]);
  assert(traces[0] == 'trace-a');
  assert(traces[1] == 'trace-b');

  final routedOtel = <Map<String, Object?>>[];
<<<<<<< HEAD
  final routedAll = <Map<String, Object?>>[];
  final routed = Logger(
    appName: 'checkout',
    transports: <LogTransport>[
      OpenTelemetryTransport(routedOtel.add),
      SupabaseTransport(routedAll.add),
    ],
  );
  await routed.info('default on');
  await routed.warn('opted out', otel: false);
  await routed.notOtel().error('logger opted out');
  await routed.notOtel().info('call opted back in', otel: true);
  assert(routedOtel.length == 2);
  assert(routedOtel.first['body'] == 'default on');
  assert(routedOtel.last['body'] == 'call opted back in');
  assert(routedAll.length == 4);
  assert(routed.otel, 'notOtel() must return a derived logger');
=======
  final regular = MemoryTransport();
  final routedLogger = Logger(
    appName: 'routing',
    otel: false,
    transports: <LogTransport>[OpenTelemetryTransport(routedOtel.add), regular],
  );
  final defaultOff = routedLogger.event(LogLevel.info, 'default-off');
  assert(!defaultOff.isOtelEnabled(routedLogger.isOtelEnabled()));
  await defaultOff.send();
  await routedLogger.event(LogLevel.info, 'forced-on').useOtel().send();
  await routedLogger
      .event(LogLevel.info, 'reset-off')
      .useOtel()
      .resetOtel()
      .send();
  routedLogger.useOtel();
  await routedLogger.event(LogLevel.warn, 'forced-off').notOtel().send();
  await routedLogger.event(LogLevel.info, 'logger-on').withOtel(true).send();
  assert(
    routedOtel.map((record) => record['body']).toList().join(',') ==
        'forced-on,logger-on',
  );
  assert(
    regular.records.map((record) => record['message']).toList().join(',') ==
        'default-off,forced-on,reset-off,forced-off,logger-on',
  );
>>>>>>> 0b2ae1c6cf9be0147ff386f3659a554c3853e666

  print('Dart/Flutter next-loggers conformance passed');
}

File: sdk/wasm/tests/conformance.rs
use next_loggers_wasm::{
    LogContext, LogLevel, Logger, OpenTelemetryTransport, SupabaseTransport, Transport, SCHEMA,
};
use std::collections::BTreeMap;
use std::sync::{Arc, Mutex};

#[test]
fn explicit_host_context_flows_to_otel_and_supabase() {
    let otel = Arc::new(Mutex::new(Vec::new()));
    let supabase = Arc::new(Mutex::new(Vec::new()));
    let otel_sink = otel.clone();
    let supabase_sink = supabase.clone();

    let logger = Logger::new("payments")
        .unwrap()
        .named("audit")
        .with_field("environment", "test")
        .with_id_factory(|| "wasm-record-1".to_string())
        .with_clock(|| "2026-01-02T03:04:05.000Z".to_string())
        .with_transport(Arc::new(OpenTelemetryTransport::new(move |record| {
            otel_sink.lock().unwrap().push(record);
            Ok(())
        })))
        .with_transport(Arc::new(SupabaseTransport::new(move |record| {
            supabase_sink.lock().unwrap().push(record);
            Ok(())
        })));

    let context = LogContext {
        trace_id: Some("0123456789abcdef0123456789abcdef".to_string()),
        span_id: Some("0123456789abcdef".to_string()),
        trace_flags: 1,
        trace_state: Some("vendor=value".to_string()),
        fields: BTreeMap::from([("requestId".to_string(), "request-1".to_string())]),
        tags: vec!["otel".to_string(), "wasm".to_string(), "otel".to_string()],
    };

    let record = logger
        .log(
            LogLevel::Error,
            "payment failed",
            Some(&context),
            BTreeMap::from([("orderId".to_string(), "order-42".to_string())]),
        )
        .unwrap();

    assert_eq!(record.schema, SCHEMA);
    assert_eq!(record.level, LogLevel::Error);
    assert_eq!(
        record.trace_id.as_deref(),
        Some("0123456789abcdef0123456789abcdef")
    );
    assert_eq!(
        record.fields.get("otel.span_id").map(String::as_str),
        Some("0123456789abcdef")
    );
    assert_eq!(
        record.fields.get("requestId").map(String::as_str),
        Some("request-1")
    );
    assert_eq!(
        record.fields.get("orderId").map(String::as_str),
        Some("order-42")
    );
    assert_eq!(record.tags, vec!["otel", "wasm"]);

    let otel = otel.lock().unwrap();
    assert_eq!(otel.len(), 1);
    assert_eq!(otel[0].severity_number, 17);
    assert_eq!(
        otel[0].attributes.get("trace.id").map(String::as_str),
        Some("0123456789abcdef0123456789abcdef")
    );
    assert_eq!(supabase.lock().unwrap().as_slice(), &[record]);
}

#[test]
fn context_is_never_hidden_in_global_or_thread_local_state() {
    let logger = Logger::new("app").unwrap();
    let first = LogContext {
        trace_id: Some("trace-a".to_string()),
        ..LogContext::default()
    };
    let second = LogContext {
        trace_id: Some("trace-b".to_string()),
        ..LogContext::default()
    };
    assert_eq!(
        logger.info("a", Some(&first)).unwrap().trace_id.as_deref(),
        Some("trace-a")
    );
    assert_eq!(
        logger.info("b", Some(&second)).unwrap().trace_id.as_deref(),
        Some("trace-b")
    );
    assert_eq!(logger.info("none", None).unwrap().trace_id, None);
}

#[test]
fn transport_errors_are_visible_to_the_host() {
    struct Broken;
    impl Transport for Broken {
        fn write(&self, _record: &next_loggers_wasm::LogRecord) -> Result<(), String> {
            Err("offline".to_string())
        }
    }

    let logger = Logger::new("app").unwrap().with_transport(Arc::new(Broken));
    assert_eq!(logger.error("boom", None).unwrap_err(), "offline");
}

#[test]
fn per_event_opentelemetry_routing_preserves_regular_transports() {
    let otel = Arc::new(Mutex::new(Vec::new()));
    let regular = Arc::new(Mutex::new(Vec::new()));
    let otel_sink = otel.clone();
    let regular_sink = regular.clone();
    let mut logger = Logger::new("app")
        .unwrap()
        .with_otel_enabled(false)
        .with_transport(Arc::new(OpenTelemetryTransport::new(move |record| {
            otel_sink.lock().unwrap().push(record);
            Ok(())
        })))
        .with_transport(Arc::new(SupabaseTransport::new(move |record| {
            regular_sink.lock().unwrap().push(record);
            Ok(())
        })));

    let default_off = logger.event(LogLevel::Info, "default-off", None, BTreeMap::new());
    assert!(!default_off.is_otel_enabled(logger.is_otel_enabled()));
    default_off.send().unwrap();
    logger
        .event(LogLevel::Info, "forced-on", None, BTreeMap::new())
        .use_otel()
        .send()
        .unwrap();
    logger
        .event(LogLevel::Info, "reset-off", None, BTreeMap::new())
        .use_otel()
        .reset_otel()
        .send()
        .unwrap();
    logger.set_otel_enabled(true);
    logger
        .event(LogLevel::Warn, "forced-off", None, BTreeMap::new())
        .not_otel()
        .send()
        .unwrap();
    logger
        .event(LogLevel::Info, "logger-on", None, BTreeMap::new())
        .with_otel(true)
        .send()
        .unwrap();

    let otel = otel.lock().unwrap();
    assert_eq!(otel.len(), 2);
    assert_eq!(otel[0].body, "forced-on");
    assert_eq!(otel[1].body, "logger-on");
    drop(otel);
    assert_eq!(
        regular
            .lock()
            .unwrap()
            .iter()
            .map(|record| record.message.as_str())
            .collect::<Vec<_>>(),
        [
            "default-off",
            "forced-on",
            "reset-off",
            "forced-off",
            "logger-on"
        ]
    );
}
